Samples/AudioCreation/cs/AudioCreation/Scenario7_RecordPlayback.xaml.cs

```diff
@@ -0,0 +1,570 @@
+
+            if (requestStartPlaying)
+            {
+                Debug.Assert(!isRecording);
+                Debug.Assert(!requestStartRecording);
+                Debug.Assert(!isPlaying);
+
+                isPlaying = true;
+                requestStartPlaying = false;
+                currentIndex = 0;
+            }
+
+            if (isPlaying)
+            {
+                Debug.Assert(!isRecording);
+                Debug.Assert(!requestStartPlaying);
+                Debug.Assert(!requestStartRecording);
+
+                HandleOutgoingAudio(requiredSamples);
+
+                /*
+                int currentStatusMsec = DateTime.Now.Millisecond;
+                // update status every tenth of a second
+                if (lastStatusMsec % 100 > currentStatusMsec % 100)
+                {
+                    CoreApplication.MainView.CoreWindow.Dispatcher.TryRunAsync(
+                        Windows.UI.Core.CoreDispatcherPriority.Normal,
+                        UpdateStatusWhilePlaying);
+                }
+                lastStatusMsec = currentStatusMsec;
+                */
+            }
+        }
+
+        // We keep a 1MB AudioFrame and reuse it.
+        static AudioFrame s_audioFrame = new Windows.Media.AudioFrame(1024 * 1024);
+        private int s_zeroByteOutgoingFrameCount;
+
+        /// <summary>
+        /// Handle a frame of outgoing audio to the output device.
+        /// </summary>
+        /// <remarks>
+        /// This pushes a megabyte of audio each time, to greatly reduce the number of callbacks.
+        /// 
+        /// We don't need the requiredSamples argument but we don't 
+        /// </remarks>
+        private unsafe void HandleOutgoingAudio(uint requiredSamples)
+        {
+            if (requiredSamples == 0)
+            {
+                s_zeroByteOutgoingFrameCount++;
+                return;
+            }
+
+            using (AudioBuffer buffer = s_audioFrame.LockBuffer(AudioBufferAccessMode.Write))
+            using (IMemoryBufferReference reference = buffer.CreateReference())
+            {
+                byte* dataInBytes;
+                uint capacityInBytes;
+
+                // Get the buffer from the AudioFrame
+                ((IMemoryBufferByteAccess)reference).GetBuffer(out dataInBytes, out capacityInBytes);
+
+                uint bytesRemaining = capacityInBytes;
+                uint position = 0;
+
+                while (bytesRemaining > 0)
+                {
+                    uint bytesInThisIteration = bytesRemaining;
+                    if (currentIndex + bytesRemaining > maxIndex)
+                    {
+                        bytesInThisIteration = maxIndex - currentIndex;
+                        if (bytesInThisIteration == 0)
+                        {
+                            // wraparound; next iteration will get more bytes
+                            currentIndex = 0;
+                            continue;
```

Comment:
If there is less than 1 MB of valid data in byteBuffer, then (maxIndex - currentIndex) will be less than bytesRemaining (since bytesRemaining is hardcoded to 1 MB). Then two different bugs can happen: 1) When you set currentIndex to 0, currentIndex may fall outside of the range that contains valid data, so you end up sending invalid data. 2) If currentIndex is already 0, we may get into an infinite loop since currentIndex + bytesRemaining can still be larger than maxIndex.